Code quality and Fix: Front end and back end about hotkey function

ChenYiLins opened the pull request:
### Description

This is a huge PR, but it doesn't actually involve any function update. The only function change is that we can edit hotkeys when the hotkey function is disabled. See it in https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/issues/702#issuecomment-3418007928. Let me briefly explain this PR:

- Change the front-end and back-end to use [`Virtual-Key`](https://learn.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes) at the same time (this is not the exclusive benefit of Windows App SDK, this comes from win32). Because the [`RegisterHotKey`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-registerhotkey) already uses `Virtual-Key`, it is not necessary to convert `Winform's Keys` into VK.
- The configuration file needs to be changed. In the past, each key was connected by `+`, but now we change it to **Decorative key code** and **primary key code**, which are connected by `,`.

Before:

```yaml
Hotkeys:
  Enabled: true
  ForceLight: Ctrl + D1
  ForceDark: 
```

After:

```yaml
Hotkeys:
  Enabled: true
  ForceLight: 2,49
  ForceDark: 
```

- So the first is the change of the back end **HotkeyHandler**. There is no need to convert `Keys` to VK in the `Register` function. We only need to split the configuration obtained by `builder`, and we will get the original `Modifiers` and `keys`. Just call the `RegisterHotKey` directly.

Before:

```c#
public class HotkeyInternal
{
    public int Id { get; set; }
    public string StringCode { get; set; }
    public List<Keys> Modifiers { get; set; } = new();
    public Keys Key { get; set; }
    public Action Action { get; set; }
}
```

After:

```c#
public class HotkeyInternal
{
    public int Id { get; set; }
    public string StringCode { get; set; }
    public uint Modifiers { get; set; }
    public uint KeyCode { get; set; }
    public Action Action { get; set; }
}
```

- Then there is the change about the **Service**, and this part will become extremely simple! The `WM_HOTKEY` message originally uses `Virtual-Key` Code: the lower 16 bits of `lParam` are decorative key marks, and the upper 16 bits are virtual keycode.
- Then the front end part was modified by MVVM, and the `HotkeysDataObject` was put into **Models**. Then the original way of reading keys is modified. Just convert the read `Virtual-Key`. (It's just one more step ( •̀ ω •́ )✧)

Code:

```c#
    private static string GetKeyDisplayName(VirtualKey key)
    {
        return key switch
        {
            VirtualKey.Enter => "Enter",
            VirtualKey.Escape => "Esc",
            VirtualKey.Space => "Space",
            VirtualKey.Back => "Backspace",
            VirtualKey.Delete => "Del",
            VirtualKey.PageUp => "PgUp",
            VirtualKey.PageDown => "PgDn",
            VirtualKey.CapitalLock => "CapsLock",
            (VirtualKey)188 => ",",
            (VirtualKey)190 => ".",
            (VirtualKey)191 => "/",
            (VirtualKey)187 => "=",
            (VirtualKey)189 => "-",
            (VirtualKey)219 => "[",
            (VirtualKey)221 => "]",
            (VirtualKey)220 => "\\",
            (VirtualKey)186 => ";",
            (VirtualKey)222 => "'",
            (VirtualKey)192 => "`",
            _ => key.ToString(),
        };
    }
```

### Screenshots

![PixPin_2025-10-18_23-31-32](https://github.com/user-attachments/assets/36aa2a25-e535-4fd4-83cf-d9b6d9ddbe94)


That's all about the explanation. I hope this is a good step to start updating the 11.X version.
At the same time, @Spiritreader @Armin2208 need you to check more. I'm worried that some minor mistakes will ruin the back end. Thank you very much!

ChenYiLins committed:
Code quality: On the optimization of MVVM

ChenYiLins committed:
Fix: Unified front end and back end use Virtual-Key

ChenYiLins committed:
Code quality: Allow hotkeys to be edited at any time

ChenYiLins committed:
Code quality: Remove info

Spiritreader commented on AutoDarkModeSvc/Service.cs:
```diff
@@ -472,24 +472,11 @@ protected override void WndProc(ref Message m)
-            var match = HotkeyHandler.GetRegistered(modifiersPressed, key);
-            if (match == null)
-            {
-                if (modifiersPressed.Contains(Keys.LWin))
```
Would your change correctly handle the case that you can press both window keys at once?

ChenYiLins replied:
I'm a little confused. shouldn't the right Win key still be displayed as Win key? (just like right Ctrl and left Ctrl). We shouldn't need to capture two key presses at the same time, just as we won't capture holding down the left and right Ctrl at the same time. And I find [it](https://superuser.com/questions/1138399/what-is-the-purpose-of-the-right-windows-key).

Spiritreader commented on AutoDarkModeSvc/Handlers/HotkeyHandler.cs:
```diff
@@ -1,216 +1,204 @@
+            var parts = hotkeyString.Split(',');
+            if (parts.Length != 2 || !uint.TryParse(parts[0], out uint modifiers) || !uint.TryParse(parts[1], out uint keyCode))
+            {
+                _logger.Error($"Invalid hotkey format: {hotkeyString}");
```
The hotkey string will map to the numbered keycodes instead of verbose key names. This makes it difficult to read / parse / troubleshoot when looking at the log file.

ChenYiLins committed:
Code quality: Converter to custom name

ChenYiLins commented on AutoDarkModeSvc/Handlers/HotkeyHandler.cs:
```diff
@@ -1,216 +1,204 @@
     {
         if (Service == null)
         {
-            Logger.Error("service not instantiated while trying to unregister hotkey");
```
Ok, I recovered in the new commit.

ChenYiLins commented:
@Spiritreader I made some changes according to your reply.

> Could you please adjust your formatter / linter to match the current format?

Ah, sorry. I'm using **CSharpier**. I can hardly control these behaviors, but nowI have stopped using it to make it easier to check.

> It makes it very hard to review this PR if there is mixed syntax and functional changes.
> 
> There's also another concern. Since the backend stores the keycodes in the config file, it will invalidate all current hotkey configurations.

Now I have added **HotkeyConverter** to Lib's **Helper** (most of the code is based on my previous commit). To put it simply, I chose to use the value of VK Codes before, and now it is written in the configuration file too, but it is more readable: 

```c#
return key.ToString();
```

Indeed, in fact, **HotkeyStringConverter** plays the role of translation.

> There is no way to tell what 2,49 is, whereas Ctrl + D1 is more humanly readable. I would like to keep the yaml configuration file verbose to a point where you can understand what's going on.

So now the configuration file has almost returned to the previous readable state, and even I think it is more friendly to the front-end and back-end, because as I said before, `RegisterHotKey` originally uses `Virtual-Key`.

> Is there any chance we might encounter some edge cases on where the hotkey conversion fails? And in worst case scenario, we still get the actual raw keycode numbers, correct?

In fact, your guess should exist. If we decompile **VirtualKey**, we can see that in C#, it is missing the key values of 173-194 and 219-254. Although some values can't be seen in our common keyboard types at all, `SpecialKeyMap` is to make up for this defect, that is to say, if we make up all these defects, edge cases will no longer exist.

More info: 
- https://github.com/microsoft/microsoft-ui-xaml/issues/2159
- https://stackoverflow.com/questions/32098242/how-to-handle-missing-values-from-virtualkey-enumeration-in-keyroutedeventargs-e

ChenYiLins committed:
Code quality: Support for old hotkeys

Armin2208 commented:
What happens, if:

1. You set the force dark mode hotkey to be ctrl + 1.
2. enable hotkeys
3. You actually want now that force light mode to be ctrl + 1, so you open the edit screen of force light mode hotkey and press ctrl + 1

Spiritreader commented:
> What happens, if:
> 
>     1. You set the force dark mode hotkey to be ctrl + 1.
> 
>     2. enable hotkeys
> 
>     3. You actually want now that force light mode to be ctrl + 1, so you open the edit screen of force light mode hotkey and press ctrl + 1

@ChenYiLins If this is an issue, we could temporarily disable the hotkey functionality by setting the config option in the yaml file to false and sving it when opening the edit dialog. Then afterwards, we could re-enable it when the user has finished configuring their hotkey.

Armin2208 commented:
> @ChenYiLins If this is an issue, we could temporarily disable the hotkey functionality by setting the config option in the yaml file to false and sving it when opening the edit dialog. Then afterwards, we could re-enable it when the user has finished configuring their hotkey.

I think this is an epic solution! But we should also beware that if the user closes the window (X or altf4) while the edit dialog is opened, it should also recover the switch hotkey ability. Because if someone clicks on the edit button, closes the window with the X, he kinda loses his hotkey abilities.

Spiritreader commented:
> > @ChenYiLins If this is an issue, we could temporarily disable the hotkey functionality by setting the config option in the yaml file to false and sving it when opening the edit dialog. Then afterwards, we could re-enable it when the user has finished configuring their hotkey.
> 
> I think this is an epic solution! But we should also beware that if the user closes the window (X or altf4) while the edit dialog is opened, it should also recover the switch hotkey ability. Because if someone clicks on the edit button, closes the window with the X, he kinda loses his hotkey abilities.

There's two ways to do this. We should discuss on which ones we want to use

## 1. Frontend only
Ensure that if we are in that state and the frontend is terminated with alt+f4 we re-enable the setting

Advantages:
- less changes
- easier to implement

Disadvantages:
- can potentially fail
- requires weird editing of the config file

## 2. Backend support for temporarily disabling hotkeys
We add a command, for example `--enter-hotkey-config-mode` and `--exit-hotkey-config-mode`
- The enter-config mode stops the hotkey listener for, let's say, 60 seconds and has to be refreshed by the frontend every time (check-in)
- The exit-config mode re-enables the hotkey functionality, but only if the feature is turned on in the config

We can then resend the check-in command every 50 seconds once the dialog is opened. If anything happens to the app, hotkeys get automatically re-enabled (if they are enabled in the config)
If the dialog is closed, we call the exit-mode function explicity, also re-enabling hotkeys (if they are enabled in the config).

With disabling hotkeys, I probably think it's a good idea to add a variable to the GlobalState that gets set if hotkeys should be temporarily disabled, and then just return in the wndproc function that receives the hotkey events.

Advantages:
- config file is not touched, can be done in-place
- never fails, automatically restores hotkey state if check-ins from the frontend stop

Disadvantages:
- more complex to implement
- requires both frontend and backend changes

ChenYiLins commented:
> > @Jay-o-Way oh okay, so the current state is broken in stable? How does the described behavior behave in this PR?
> 
> Mostly the same (updated earlier comment too) Also noticed that pressing Win+Shift+1 acts like Win+1, which opens the first pinned app on the taskbar. Hotkeys are extremely difficult to get right.

Yes, I also found this problem, which existed in the previous 10 versions of ADM. But I think we can actually try to improve the way to get hot keys to solve this problem. (This is absolutely possible)

ChenYiLins commented:
> @ChenYiLins If this is an issue, we could temporarily disable the hotkey functionality by setting the config option in the yaml file to false and sving it when opening the edit dialog. Then afterwards, we could re-enable it when the user has finished configuring their hotkey.

Great! (Why didn't we think of before hahah)

martinwilco committed:
Translated using Weblate (German)

Currently translated at 100.0% (288 of 288 strings)

Co-authored-by: Martin Wilichowski <martin.wilichowski@gmx.de>
Translate-URL: https://hosted.weblate.org/projects/auto-dark-mode/app/de/
Translation: Auto Dark Mode/App

Commit:
Translated using Weblate (Chinese (Traditional Han script))

Currently translated at 99.3% (286 of 288 strings)

Co-authored-by: hugoalh <hugoalh@users.noreply.hosted.weblate.org>
Translate-URL: https://hosted.weblate.org/projects/auto-dark-mode/app/zh_Hant/
Translation: Auto Dark Mode/App

mojienjoyment committed:
Translated using Weblate (Persian)

Currently translated at 86.1% (248 of 288 strings)

Co-authored-by: Moji <moj.enjoy@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/auto-dark-mode/app/fa/
Translation: Auto Dark Mode/App

kazutoiris committed:
Translated using Weblate (Chinese (Simplified Han script))

Currently translated at 100.0% (288 of 288 strings)

Translated using Weblate (Chinese (Simplified Han script))

Currently translated at 100.0% (288 of 288 strings)

Co-authored-by: Kazuto Iris <78157415+kazutoiris@users.noreply.github.com>
Translate-URL: https://hosted.weblate.org/projects/auto-dark-mode/app/zh_Hans/
Translation: Auto Dark Mode/App

DanielSevillano committed:
Translated using Weblate (Spanish)

Currently translated at 84.3% (243 of 288 strings)

Co-authored-by: Daniel <daniel-sevillano@hotmail.com>
Translate-URL: https://hosted.weblate.org/projects/auto-dark-mode/app/es/
Translation: Auto Dark Mode/App

turi1122 committed:
Translated using Weblate (Hungarian)

Currently translated at 98.2% (283 of 288 strings)

Co-authored-by: turi1122 <61917421+turi1122@users.noreply.github.com>
Translate-URL: https://hosted.weblate.org/projects/auto-dark-mode/app/hu/
Translation: Auto Dark Mode/App

ChenYiLins commented:
> ## 2. Backend support for temporarily disabling hotkeys

Let me try to see if there is a simpler solution. (ง •_•)ง

ChenYiLins committed:
Code quality: Change hotkey acquisition to system hook

ChenYiLins committed:
Merge branch 'v11.1' of https://github.com/AutoDarkMode/Windows-Auto-Night-Mode into hotkeyspage

# Conflicts:
#	AutoDarkModeApp/Strings/zh-hans/Resources.resw

ChenYiLins committed:
Merge branch 'weblate-auto-dark-mode-app' of https://github.com/weblate/Windows-Auto-Night-Mode into hotkeyspage

# Conflicts:
#	AutoDarkModeApp/Strings/fa/Resources.resw
#	AutoDarkModeApp/Strings/hu/Resources.resw
#	AutoDarkModeApp/Strings/zh-hans/Resources.resw
#	AutoDarkModeApp/Strings/zh-hant/Resources.resw

ChenYiLins committed:
Fix: Strings error

ChenYiLins committed:
Merge branch 'v11.1' of https://github.com/AutoDarkMode/Windows-Auto-Night-Mode into hotkeyspage

# Conflicts:
#	AutoDarkModeApp/Strings/hu/Resources.resw
#	AutoDarkModeApp/Strings/zh-hant/Resources.resw

ChenYiLins committed:
Merge branch 'weblate-auto-dark-mode-app' of https://github.com/weblate/Windows-Auto-Night-Mode into hotkeyspage

# Conflicts:
#	AutoDarkModeApp/Strings/hu/Resources.resw
#	AutoDarkModeApp/Strings/zh-hans/Resources.resw
#	AutoDarkModeApp/Strings/zh-hant/Resources.resw
